src/ipa_pc/mod.rs

```diff
@@ -744,55 +784,81 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
+                }
 
-                combined_polynomial += (cur_challenge, polynomial);
-                combined_commitment_proj += &commitment.comm[0].mul(cur_challenge);
-            }
+                cur_challenge = opening_challenges(opening_challenge_counter);
+                opening_challenge_counter += 1;
 
-            if hiding_bound.is_some() {
-                has_hiding = true;
-                combined_rand += &(cur_challenge * &randomness.randomness().rand);
-            }
+                if let Some(degree_bound_len) = degree_bound_len {
+                    let shifted_degree_bound = degree_bound_len % key_len - 1;
+                    let last_segment_polynomial = Polynomial::from_coefficients_slice(&polynomial.coeffs[(segments_count - 1) * key_len..p_len]);
+                    let shifted_polynomial = Self::shift_polynomial(
+                        ck,
+                        &last_segment_polynomial,
+                        shifted_degree_bound
+                    );
+                    let shift = -point.pow(&[(key_len - shifted_degree_bound - 1) as u64]);
 
-            cur_challenge = opening_challenges(opening_challenge_counter);
-            opening_challenge_counter += 1;
+                    combined_polynomial += (cur_challenge, &shifted_polynomial);
+                    combined_commitment_proj += &commitment.shifted_comm.unwrap().mul(cur_challenge);
 
-            let d = degree_bound.and_then(|d| {
-                if (d + 1) % key_len != 0 { Some(d + 1) } else { None }
-            });
+                    combined_polynomial += (cur_challenge * &shift, &last_segment_polynomial);
+                    combined_commitment_proj += &commitment.comm[segments_count - 1].mul(cur_challenge * &shift);
 
-            assert_eq!(
-                d.is_some(),
-                commitment.shifted_comm.is_some(),
-                "shifted_comm mismatch for {}",
-                label
-            );
+                    if hiding_bound.is_some() {
+                        let shifted_rand = labeled_randomness.randomness().shifted_rand;
+                        assert!(
+                            shifted_rand.is_some(),
+                            "shifted_rand.is_none() for {}",
+                            label
+                        );
+                        combined_rand += &(cur_challenge * &shifted_rand.unwrap());
+                        combined_rand += &(cur_challenge * &shift * &labeled_randomness.randomness().rand[segments_count - 1]);
+                    }
 
-            assert_eq!(
-                degree_bound,
-                labeled_commitment.degree_bound(),
-                "labeled_comm degree bound mismatch for {}",
-                label
-            );
+                    cur_challenge = opening_challenges(opening_challenge_counter);
+                    opening_challenge_counter += 1;
+                }
+
+            } else {
```

Comment:
doesn't the code in the if clause work also for the case that segments_count = 1?

Reply:
I've refactored this part a bit to avoid unnecessary code duplication between segmented and non-segmented parts